worker/changestream/eventqueue/eventqueue.go

```diff
@@ -0,0 +1,247 @@
+	s.subscriptionsCount++
+	sub := &subscription{
+		id:           s.subscriptionsCount,
+		changes:      make(chan changestream.ChangeEvent),
+		topics:       make(map[string]struct{}),
+		unsubscriber: func() { s.unsubscribe(s.subscriptionsCount) },
+	}
+	s.subscriptions[sub.id] = sub
+
+	// Register filters to route changes matching the subscription criteria to the newly crated subscription.
+	for _, opt := range opts {
+		namespace := opt.Namespace()
+		s.subscriptionsByNS[namespace] = append(s.subscriptionsByNS[namespace], &eventFilter{
+			subscriptionID: sub.id,
+			changeMask:     opt.ChangeMask(),
+			filter:         opt.Filter(),
+		})
+		sub.topics[namespace] = struct{}{}
+	}
+
+	return sub, nil
+}
+
+// Kill stops the event queue.
+func (s *EventQueue) Kill() {
+	s.tomb.Kill(nil)
+}
+
+// Wait waits for the event queue to stop.
+func (s *EventQueue) Wait() error {
+	return s.tomb.Wait()
+}
+
+func (s *EventQueue) unsubscribe(subscriptionID int) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	sub, found := s.subscriptions[subscriptionID]
+	if !found {
+		return
+	}
+
+	for topic := range sub.topics {
+		var updatedFilters []*eventFilter
+		for _, filter := range s.subscriptionsByNS[topic] {
+			if filter.subscriptionID == subscriptionID {
+				continue
+			}
+			updatedFilters = append(updatedFilters, filter)
+		}
+		s.subscriptionsByNS[topic] = updatedFilters
+	}
+
+	delete(s.subscriptions, subscriptionID)
+
+	// Close the subscription channel after a dispatch. If a unsubscription
+	// happens during a dispatch, then it clauses the event queue to potentially
+	// panic. Removal of the subscription from the queue is synchronous, so
+	// the subscription can no longer be used when another dispatch occurs. The
+	// closing of the channel then becomes asynchronous and can be closed during
+	// the next selection.
+	s.tomb.Go(func() error {
+		// Ensure that we give up if the tomb is dying.
+		select {
+		case <-s.tomb.Dying():
+			return tomb.ErrDying
+		case s.actions <- func() {
+			close(sub.changes)
+		}:
+		}
+		return nil
+	})
+}
+
+func (s *EventQueue) loop() error {
+	defer func() {
+		s.mutex.Lock()
+		defer s.mutex.Unlock()
+
+		for _, sub := range s.subscriptions {
+			close(sub.changes)
+		}
+		s.subscriptions = make(map[int]*subscription)
+	}()
+
+	for {
+		var (
+			ch changestream.ChangeEvent
+			ok bool
+		)
+		select {
+		case <-s.tomb.Dying():
+			return tomb.ErrDying
+		case action := <-s.actions:
+			action()
+			continue
+		case ch, ok = <-s.stream.Changes():
+			// If the stream is closed, we expect that a new worker will come
+			// again using the change stream worker infrastructure. In this case
+			// just ignore and close out.
+			if !ok {
+				s.logger.Infof("change stream change channel is closed")
+				return nil
+			}
+		}
+
+		// The dispatch is done in two stages to avoid holding the lock for the
+		// duration of the dispatch.
+		subs := s.gatherSubscriptions(ch)
```

Comment:
Why not do this dispatch inside the `s.stream.Changes` case?

Reply:
Because we will hold the lock for the subscriptions, which then causes a dead lock if attempting to unsubscribe at the same time. I'm currently revisiting how we unsubscribe atm. That's the reason for the two different methods.